lib/libzfs/libzfs_sendrecv.c

```diff
@@ -3439,7 +3551,16 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 		goto out;
 	}
 
-	err = ioctl_err = lzc_receive_one(destsnap, props, origin,
+	toplevel = chopprefix[0] != '/';
+	if (drrb->drr_type == DMU_OST_ZVOL)
+		type = ZFS_TYPE_VOLUME;
+	else
```

Comment:
can we be a bit more defensive here?
I'm thinking something like
else if (drrb->drr_type == DMU_OST_ZFS) type = ZFS_TYPE_FILESYSTEM; else goto out;
or something simular?